### PR TITLE
Close nodes properly in Coordinator tests

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today closing a `ClusterNode` in an `AbstractCoordinatorTestCase` uses
`onNode()` so has no effect if the node is not in the current list of nodes.
It also discards the `Runnable` it creates without having run it, so has no
effect anyway.

This commit makes these tests much stricter about properly closing the nodes
started during `Coordinator` tests, by tracking the persisted states that are
opened, and adds an assertion to catch the trappy requirement that the closing
node still belongs to the cluster.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I strongly recommend looking at this with whitespace changes hidden. Rather than +1007/-901 it's more like +175/-69.

File: test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java [hunk 1]
```diff
@@ -903,15 +931,16 @@ protected ConnectionStatus getConnectionStatus(DiscoveryNode destination) {
...
-                    logger.trace("taking down [{}]", localNode);
+                    logger.trace("closing");
                     coordinator.stop();
                     clusterService.stop();
                     //transportService.stop(); // does blocking stuff :/
                     clusterService.close();
                     coordinator.close();
                     //transportService.close(); // does blocking stuff :/
-                });
+                }).run();
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java]: oops 👿

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java [hunk 2]
```diff
@@ -1096,41 +1151,43 @@ public void writeTo(StreamOutput out) throws IOException {
...
             return builder;
         }
+
     }
 
     public void testClusterRecoversAfterExceptionDuringSerialization() {
-        final Cluster cluster = new Cluster(randomIntBetween(2, 5)); // 1-node cluster doesn't do any serialization
-        cluster.runRandomly();
-        cluster.stabilise();
+        try (Cluster cluster = new Cluster(randomIntBetween(2, 5))) { // 1-node cluster doesn't do any serialization
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: this is not true anymore. Perhaps we should amend this test.

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: Good point, this is fine for 1-node clusters too now. Addressed in 92a02b7.

**Pull request merged**